File: Project4/main.py
"""
File: main.py
Professor: Professor Kapleau
Course: CS280
Section: 011

This program will create a test a Complex numbers class which performs standard arithmetic, 
other extra functions, and returns useful information about the complex number. 
Explanatory comments are provided below.

Note: the output is rounded to stop floating point arithmetic issues.
"""

#kFLFBx9T7XJmrr3p2F/fmAj+8DS8GqnKeq/TOIAPcuI=

import math
class Complex:
    # constructor
    def __init__(self, real = 0.0, img = 0.0):
        self.real = real
        self.img = img

    # returns the real component
    def re(self):
        return self.real
    
    # returns the imaginary component
    def im(self):
        return self.img
    
    # returns the phase angle of the complex number
    def phase(self):
        return math.atan2(self.img, self.real)
    
    # returns the modulus or magnitude of the complex number
    def magnitude(self):
        return math.hypot(self.real, self.img)
    
    # returns the conjugate of the complex number
    def conjugate(self):
        return Complex(self.real, -1.0*self.img)
    
    # returns the complex number that represents the (c + c) or (c + d) addition 
    def __add__(self, t):
        if isinstance(t, Complex):
            return Complex(self.real + t.real, self.img + t.img)
        elif isinstance(t, (float, int)):
                return Complex(self.real + t, self.img)
        else:
            raise TypeError
    
    # returns the complex number that represents the (d + c) addition
    def __radd__(self, d):
        if isinstance(d, (float, int)):
            return self + d
        else:
            raise TypeError

    # returns the complex number that represents the (c - c) or (c - d) subtraction
    def __sub__(self, c):
        if isinstance(c, Complex):
            return Complex(self.real - c.real, self.img - c.img)
        elif isinstance(c, (float, int)):
            return Complex(self.real - c, self.img)
        else:
            raise TypeError
    
    # returns the complex number that represents the (d - c) subtraction
    def __rsub__(self, d):
        if isinstance(d, (float, int)):
            return -1.0 * self + d
        else:
            raise TypeError

    # returns the complex number that represents the (c * c) or (c * d) multiplication
    def __mul__(self, t):
        if isinstance(t, Complex):
            return Complex(self.real*t.real - self.img*t.img, self.real*t.img + self.img*t.real)
        elif isinstance (t, (float, int)):
            return Complex(t*self.real, t*self.img)
        else:
            raise TypeError
    
    # returns the complex number that represents the (d * c) multiplication
    def __rmul__(self, d):
        if isinstance(d, (float, int)):
            return self * d
        else:
            raise TypeError
    
    # returns the complex number that represents the (c / c) or (c / d) division
    def __truediv__(self, t):
        if isinstance(t, Complex):
            mag = self.magnitude() / t.magnitude()
            theta = self.phase() - t.phase()
            return Complex(mag*math.cos(theta), mag*math.sin(theta))    
        elif isinstance (t, (float, int)):
            return Complex(self.real/ t, self.img/ t)
        else:
            raise TypeError
        
    # returns the complex number that represents the (d / c) division
    def __rtruediv__(self, d):
        if isinstance(d, (float, int)):
            return d * Complex(round(math.cos(-1*self.phase())/self.magnitude(), 3), round(math.sin(-1*self.phase())/self.magnitude(), 3))
        else:
            raise TypeError
    

    # returns true or false depending on whether they represent the same complex number
    def __eq__(self, c):
        if isinstance(c, Complex):
            return self.real == c.real and self.img == c.img
        else:
            raise TypeError

    # returns a string representation of complex numbers in algebraic form
    def __str__(self):
        return "(" + str(float(self.real)) + (" + " if self.img >= 0 else " - ") + str(float(abs(self.img))) + "i)"

if __name__ == "__main__":
    c1 = Complex(1,1)
    c2 = Complex(2,2)
    d = 2.0

    #Retrieves properties of the complex numbers
    print("Z1 = %s" % c1)
    print("Re(Z1): %f" % c1.re())
    print("Im(Z1): %f" % c1.im())
    print("Phase angle of Z1: %f" % c1.phase())
    print("Modulus or |Z1|: %f\n" % c1.magnitude())
    print("Z2 = %s" % c2)
    print("Re(Z2): %f" % c2.re())
    print("Im(Z2): %f" % c2.im())
    print("Phase angle of Z2: %f" % c2.phase())
    print("Modulus or |Z2|: %f\n" % c2.magnitude())

    #Performs standard operations on the complex numbers
    print('%s + %s = %s' % (c1, c2, c1 + c2)) #addition
    print('%s + %f = %s' % (c1, d, c1 + d)) #real number addition 1
    print('%f + %s = %s' % (d, c1, d + c1)) #real number addition 2
    print('%s - %s = %s' % (c1, c2, c1 - c2)) #subtraction
    print('%s - %f = %s' % (c1, d, c1 - d)) #real number subtraction 1
    print('%f - %s = %s' % (d, c1, d - c1)) #real number subtraction 2
    print('%s * %s = %s' % (c1, c2, c1 * c2)) #multiplication
    print('%s * %f = %s' % (c1, d, c1 * d)) #real number multiplication 1
    print('%f * %s = %s' % (d, c1, d * c1)) #real number multiplication 2
    print('%s / %s = %s' % (c1, c2, c1 / c2)) #division
    print('%s / %f = %s' % (c1, d, c1 / d)) #real number division 1
    print('%f / %s = %s\n' % (d, c1, d / c1)) #real number division 2

    #Extra features
    print("Conjugate of Z1 is: %s" % c1.conjugate())
    print("Reciprocal of Z2 is: %s" % (1 / c2))
    print("Z1 == Z2: %s" % (c1 == c2))
    print("Z1 * %f = %s" % (d, (d * c1)))
    print("(%s * Z1) == Z2: %s" % (d, ((d * c1) == c2)))

File: Project4/main.cc
/** File: main.cc
 *  Professor: Professor Kapleau
 *  Course: CS280
 *  Section: 011
 *
 *  This program will create a test a Complex numbers class which performs standard arithmetic, 
 *  other extra functions, and returns useful information about the complex number. 
 *  Explanatory comments are provided below.
 */

//kFLFBx9T7XJmrr3p2F/fmAj+8DS8GqnKeq/TOIAPcuI=

#include <iostream>
#include <cmath>
#include <cstdio>
using namespace std;

class Complex{
    private:
        //instance variables
        double real;
        double img;
    
    public:
        //constructor
        Complex(double real = 0.0, double img = 0.0) : real(real), img(img) {}
    
        //returns the real component
        double re() const{
            return real;
        }

        //returns the imaginary component
        double im() const{
            return img;
        }
        
        //returns the phase angle of the complex number
        double phase() const{
            return atan(img * 1.0 / real);
        }

        //returns the modulus or magnitude of the complex number
        double magnitude() const{
            return hypot(real, img);
        }
      
        //returns the conjugate of the complex number
        Complex conjugate(){
            return Complex(real, -img);
        }

        //returns the complex number that represents the (c + c) addition
        Complex operator+(const Complex &c) const{
            return Complex(real + c.real, img + c.img);
        }

        //returns the complex number that represents the (c + d) addition
        Complex operator+(const double &d) const{
            return Complex(real + d, img);
        }

        //returns the complex number that represents the (d + c) addition
        friend Complex operator+(double d, const Complex &c){
            return c + d;
        }

        //returns the complex number that represents the (c - c) subtraction
        Complex operator-(const Complex &c) const{
            return Complex(real - c.real, img - c.img);
        }

        //returns the complex number that represents the (c - d) subtraction
        Complex operator-(const double &d) const{
            return Complex(real - d, img);
        }

        //returns the complex number that represents the (d - c) subtraction
        friend Complex operator-(double d, const Complex &c){
            return -1 * c + d;
        }

        //returns the complex number that represents the (c * c) multiplication
        Complex operator*(const Complex &c) const{
            return Complex(real*c.real - img*c.img, real*c.img + img*c.real);
        }

        //returns the complex number that represents the (c * d) multiplication
        Complex operator*(double d) const{
            return Complex(d*real, d*img);
        }

        //returns the complex number that represents the (d * c) multiplication
        friend Complex operator*(double d, const Complex &c){
            return c * d;
        }

        //returns the complex number that represents the (c / c) division
        Complex operator/(const Complex &c) const{
            double mag = magnitude() / c.magnitude();
            double theta = phase() - c.phase();
            return Complex(mag*cos(theta), mag*sin(theta));
        }

        //returns the complex number that represents the (c / d) division
        Complex operator/(double d) const{
            return Complex(real/d, img/d);
        }

        //returns the complex number that represents the (d / c) division
        friend Complex operator/(double d, const Complex &c){ 
            return d * Complex(cos(-1*c.phase())/c.magnitude(), sin(-1*c.phase())/c.magnitude());
        }

        //returns true or false depending on whether they represent the same complex number
        bool operator == (const Complex &c){
            return real == c.real && img == c.img;
        }

        //overloads the stream insertion operator to print complex numbers in algebraic form
        friend ostream &operator<<(ostream &out, const Complex &c){
            out << '(' << c.real << (c.img >= 0 ? " + " : " - ") << abs(c.img) << "i)";
            return out;
        }
};

int main(void){

    Complex c1(1, 1);
    Complex c2(2, 2);
    double d = 2.0;

    //Retrieves properties of the complex numbers
    cout << "Z1 = " << c1 << endl;
    cout << "Re(Z1): " << c1.re() << endl;
    cout << "Im(Z1): " << c1.im() << endl;
    cout << "Phase angle of Z1: " << c1.phase() << endl;
    cout << "Modulus or |Z1|: " << c1.magnitude() << endl << endl;
    cout << "Z2 = " << c2 << endl;
    cout << "Re(Z2): " << c2.re() << endl;
    cout << "Im(Z2): " << c2.im() << endl;
    cout << "Phase angle of Z2: " << c2.phase() << endl;
    cout << "Modulus or |Z2|: " << c2.magnitude() << endl << endl;

    //Performs standard operations on the complex numbers
    cout << c1 << " + " << c2 << " = " << c1 + c2 << endl; //addition
    cout << c1 << " + " << d << " = " << c1 + d << endl; //real number addition 1
    cout << d << " + " << c1 << " = " << d + c1 << endl; //real number addition 2
    cout << c1 << " - " << c2 << " = " << c1 - c2 << endl; //subtraction
    cout << c1 << " - " << d << " = " << c1 - d << endl; //real number subtraction 1
    cout << d << " - " << c1 << " = " << d - c1 << endl; //real number subtraction 2
    cout << c1 << " * " << c2 << " = " << c1 * c2 << endl; //multiplication
    cout << c1 << " * " << d << " = " << c1 * d << endl; //real number multiplication 1
    cout << d << " * " << c1 << " = " << d * c1 << endl; //real number multiplication 2
    cout << c1 << " / " << c2 << " = " << c1 / c2 << endl; //division
    cout << c1 << " / " << d << " = " << c1 / d << endl; //real number division 1
    cout << d << " / " << c1 << " = " << d / c1 << endl << endl; //real number division 2

    //Extra features
    cout << "Conjugate of Z1 is: " << c1.conjugate() << endl;
    cout << "Reciprocal of Z2 is: " << 1 / c2 << endl;
    cout << "Z1 == Z2: " << ((c1 == c2) == 1 ? "true" : "false") << endl;
    cout << d << " * Z1" << " = " << (d * c1) << endl;
    cout << "(Z1 * " << d << ") == Z2: " << (((d * c1) == c2) == 1 ? "true" : "false") << endl;

    return 0;
}
